tests: Add tests to validate Prometheus (deployed alerting rules == default alerting rules)

Ebaneck opened the pull request:
**Component**:

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->
'tests', 'basic alerting', 'tooling'

**Context**: 

The following test is added to ensure that we can rightfully validate the deployed Prometheus alerting rules against the defaults at all times. 

See the issue here: #2538 

**Summary**:

Follow suit of PR #2542 as promised.
The guard (Heimdall the Gatekeeper) to catch Prometheus-operator chart upgrades that bring new 
undocumented alerting rules is in place.

**Acceptance criteria**: 

- Rightfully tested!
- Green build

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #ISSUE_NUMBER

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

Ebaneck committed:
test: validate that the deployed Prometheus alerting rules match the default alerting rules

We need to be sure that the deployed Prometheus alerting rules match the
default alerting rules at all times.

This test can typically fail if there is a recent Prometheus-operator chart
upgrade that brings along  new Prometheus alerting rules.

We need to be notified when alerting rules differ(when this test fails) so that Developers
can follow the new procedure to update the default rules and include the required
documentation for the newly added alerting rule

Ebaneck committed:
eve: add `alert_rules.json` to the list of test sources to be copied

To successfully compare the deployed rules against the default rules,
specifically in a multinode eve stage, we need to copy the `/tools` directory
to the bastion along side other test sources.

bert-e commented:
# Hello ebaneck,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/add-test-for-default-prometheus-rules` into
`development/2.5`
* [w/2.6/improvement/add-test-for-default-prometheus-rules](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/add-test-for-default-prometheus-rules) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote

bert-e commented:
# History mismatch



Merge commit #7c64d40741e0126d166e5ee8c1d3fc5778b65401 on the integration branch
`w/2.6/improvement/add-test-for-default-prometheus-rules` is merging a branch which is neither the current
branch `improvement/add-test-for-default-prometheus-rules` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `improvement/add-test-for-default-prometheus-rules` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #191e634fce90e777f6448a89b062f526757d6eef on the integration branch
`w/2.6/improvement/add-test-for-default-prometheus-rules` is merging a branch which is neither the current
branch `improvement/add-test-for-default-prometheus-rules` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `improvement/add-test-for-default-prometheus-rules` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard committed:
tests: add a wrapper around prometheus API

Add a class wrapping the call to Prometheus
API and add a fixture around that.

(cherry picked from commit 0acbc644178bb42115cbff254ea8fa7b41056454)

gdemonet commented on tests/post/features/monitoring.feature:
```diff
@@ -38,3 +38,8 @@ Feature: Monitoring is up and running
+    Scenario: Monitor deployed Prometheus rules
+        Given the Kubernetes API is available
+        And we have 1 running pod labeled 'prometheus=prometheus-operator-prometheus' in namespace 'metalk8s-monitoring'
+        Then the deployed Prometheus alerting rules is the same as the default alerting rules
```
```suggestion
        Then the deployed Prometheus alert rules are the same as the default alert rules
```

gdemonet commented on tests/post/steps/test_monitoring.py:
```diff
@@ -1,6 +1,8 @@
-import requests
 import json
+import pathlib
+import requests
```
Since you're re-ordering imports, `requests` is a third party library, as `pytest` and `kubernetes` are. The usual structure is the following:

```python
# Stdlib imports
import json
import pathlib

# 3rd party imports
import kubernetes.client
import pytests
import requests

# Local imports
from tests import kube_utils
from tests import utils
```

gdemonet commented on tests/post/features/monitoring.feature:
```diff
@@ -38,3 +38,8 @@ Feature: Monitoring is up and running
         Given the Kubernetes API is available
         Then a node with label 'node-role.kubernetes.io/bootstrap=' has metrics
+
+    Scenario: Monitor deployed Prometheus rules
```
I don't understand what "monitor rules" means...

```suggestion
    Scenario: Ensure deployed Prometheus rules match the default
```

gdemonet commented on tests/post/steps/test_monitoring.py:
```diff
@@ -174,6 +182,69 @@ def _node_has_metrics():
+@then(
+    "the deployed Prometheus alerting rules is the same as the default "
+    "alerting rules")
+def monitor_deployed_rules(host):
```
Same comment, this is not about monitoring rules...

```suggestion
def check_deployed_rules(host):
```

gdemonet commented on tests/post/steps/test_monitoring.py:
```diff
@@ -174,6 +182,69 @@ def _node_has_metrics():
+        When this happens, Developers need to regenerate the default alerting
+        rules using the rule_extractor python script, compare and document the
+        new changes in the Cluster monitoring guide before moving on.
+    """
```
Please comply with [PEP 257](https://www.python.org/dev/peps/pep-0257/), so we ensure consistency across our project. And if you find it hard to remember, maybe add those files to the `lint` stage in CI :wink:?

gdemonet commented on tests/post/steps/test_monitoring.py:
```diff
@@ -174,6 +182,69 @@ def _node_has_metrics():
+            "Expect status code <200> but got {} for route rules".format(
+                response.status_code
+            )
+        deployed_rules = response.json()
```
If you wait a bit for #2539, there will be a `PrometheusAPI` helper client to do this for you ;)

gdemonet commented on tests/post/steps/test_monitoring.py:
```diff
@@ -174,6 +182,69 @@ def _node_has_metrics():
+                        'name': rule['name'],
+                        'severity': rule['labels']['severity'],
+                        'message': message
+                    }
```
Is this the only info we save in the extracted `alerting_rules.json`? If yes, we can't detect if labels changed, if the expr changed... Not super useful like this, but it should at least help us detect when alerts are added/removed.

Ebaneck replied:
We have decided to also compare the `expr` as you have suggested and as such we have updated the alerting_rules.json to have this field.

For the label key in general, we have just `severity` as the subkey and yes, we do compare the severity. Maybe it would have been better to dump the label structure without altering it.

gdemonet commented on tests/post/steps/test_monitoring.py:
```diff
@@ -174,6 +182,69 @@ def _node_has_metrics():
+            )
+        deployed_rules = response.json()
+
+        # we cannot simply compare the obtained value because it contains
+        # alerts with an "activeAt": "2020-05-15T09:23:15.279979961Z",
+        # so lets filter and use only certain alerting rule fields
```
In the general sense, when you want to ensure two datasets are the same except for a few fields, don't select a subset of the fields you care about. Just ignore the fields you don't care about when comparing.

Ebaneck replied:
ok

gdemonet commented on tests/post/steps/test_monitoring.py:
```diff
@@ -174,6 +182,69 @@ def _node_has_metrics():
+
+        filename = 'alerting_rules.json'
+        filepath = (pathlib.Path(__file__)/'..'/'..'/'..'/'..'
+                    /'tools'/'rule_extractor'/filename).resolve()
```
Let's make this a constant outside of the test function.

gdemonet commented on tests/post/steps/test_monitoring.py:
```diff
@@ -174,6 +182,69 @@ def _node_has_metrics():
+        try:
+            with open(filepath) as f:
+                default_alerting_rules = json.load(f)
+        except IOError as exc:
```
Would also need to catch JSON errors ;)

gdemonet commented on tests/post/steps/test_monitoring.py:
```diff
@@ -174,6 +182,69 @@ def _node_has_metrics():
+        except IOError as exc:
+            pytest.fail("Failed to open file {}: {!s}".format(filename, exc))
+
+        assert json.dumps(default_alerting_rules, sort_keys=True) \
+            == json.dumps(deployed_alerting_rules, sort_keys=True), \
```
That works, but I'd suggest you use Python dicts equality, `pytest` automatically displays the diff when the assertion fails.

If you really want to compare a sorted list, you can do:
```python
sorted(
    default_alerting_rules,
    key=lambda rule: hash(json.dumps(rule, sort_keys=True))
)
```

But that's way more complex than required. Instead, just build a hierarchy of group/alerts and compare the final dict:

```python
# this is an example, adapt it to your code
deployed_rules = {
    group['name']: {rule['name']: rule for rule in group.rules}
    for group in prom_data['data']['groups']
}
```

alexandre-allard commented on tests/post/steps/test_monitoring.py:
```diff
@@ -1,17 +1,24 @@
 
+# Constants {{{
 
+FILE_NAME = 'alerting_rules.json'
+FILE_PATH = (pathlib.Path(__file__)/'..'/'..'/'..'/'..'
+            /'tools'/'rule_extractor'/FILE_NAME).resolve()
```
Could we use more explicit names here ?

Ebaneck replied:
Done

Ebaneck commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.6/improvement/add-test-for-default-prometheus-rules` with contents from `improvement/add-test-for-default-prometheus-rules`
and `development/2.6`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.6/improvement/add-test-for-default-prometheus-rules origin/development/2.6
 $ git merge origin/improvement/add-test-for-default-prometheus-rules
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.6/improvement/add-test-for-default-prometheus-rules
```

alexandre-allard commented on tests/post/steps/test_monitoring.py:
```diff
@@ -179,4 +192,63 @@ def _node_has_metrics():
+            "Expected default Prometheus rules to be equal to deployed rules."
+        )
+
+    utils.retry(
+        _wait_prom_rule_cmp,
+        times=5,
+        wait=3,
+        name="checking deployed Prometheus rules are equal to default rules"
+    )
```
Not sure we need to implement a retry here since we do not change anything in the step before, just reading the API, the PrometheusAPI wrapper will take care of retrying in case of non 200 exit code.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

alexandre-allard commented on tests/post/steps/test_monitoring.py:
```diff
@@ -179,4 +192,63 @@ def _node_has_metrics():
+
+        if not deployed_rules:
+            pytest.fail(
+                "Expected default deployed Prometheus rules but got {}"
+                .format(deployed_rules)
```
Don't see the point of showing the content of `deployed_rules`, with the `if` before, we know its empty.
Moreover, I'm not even sure that this case can happen since it's the return from the API in case of success, so you should at least have the status and the `data` key.
For me, you can remove this whole `if` and let the assertion fails at the bottom of this function.

bert-e commented:
# History mismatch



Merge commit #178fde1b7808f906c8f9998ce218d528d364d181 on the integration branch
`w/2.6/improvement/add-test-for-default-prometheus-rules` is merging a branch which is neither the current
branch `improvement/add-test-for-default-prometheus-rules` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `improvement/add-test-for-default-prometheus-rules` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Ebaneck commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.6/improvement/add-test-for-default-prometheus-rules` with contents from `improvement/add-test-for-default-prometheus-rules`
and `development/2.6`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.6/improvement/add-test-for-default-prometheus-rules origin/development/2.6
 $ git merge origin/improvement/add-test-for-default-prometheus-rules
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.6/improvement/add-test-for-default-prometheus-rules
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

Ebaneck commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet







*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/e520b4a80d65692a7121517004112feadd58fada) did not succeed in branch w/2.6/improvement/add-test-for-default-prometheus-rules.




*The following options are set:* **approve**

Ebaneck commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.6/improvement/add-test-for-default-prometheus-rules` with contents from `improvement/add-test-for-default-prometheus-rules`
and `development/2.6`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.6/improvement/add-test-for-default-prometheus-rules origin/development/2.6
 $ git merge origin/improvement/add-test-for-default-prometheus-rules
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.6/improvement/add-test-for-default-prometheus-rules
```





*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



Please check the status of the associated issue None.

Goodbye ebaneck.
